docs/composite-clusters.md

```diff
@@ -0,0 +1,129 @@
+
+##### Pyramid pinning
+
+In a cylinder cluster, pin requests to the overcluster are tracked by all the underclusters below. Pin requests to the underclusters, are not tracked by the overcluster, but everything is eventually pinned in the ipfs daemons at the bottom. This allows to build pinning hierarchies, where the bottom clusters have visibility to larger pinsets than the top clusters.
+
+This is a specific case of **inverse multi-ring topology** described below, with a single overcluster.
+
+### Multiple clusters have their cluster peers attached to peers from a single cluster (inverse multi-ring)
+
+In this topology, we have several overclusters which all share the same undercluster.
+
+#### Uses
+
+##### Data partitioning
+
+A group of parties may which to share a common cluster. With an inverse multi-ring setup, each party will mantain their own pinset in their own cluster, but it will use a common infrastructure (the undercluster), which is shared with others. Each overcluster may administer how cluster APIs are exposed and what authentication to use.
+
+In this case, the undercluster needs to be able to scale to be able to handle the pinsets from all the overclusters. The overclusters are probably small, with several nodes just for redundancy, connected to several of the undercluster peers.
+
+
+### Each cluster peer attached to peers from its own cluster (inception clusters)
+
+In this case, each cluster peers uses as ipfs daemon another cluster peer from the same cluster. overcluster == undercluster.
+
+#### Uses
+
+##### Destroy the world
+
+If the overcluster asks one of the peers to pin, it would in turn ask the overcluster to pin (through a different peer) and so on, ad-infinitum or until the requests falls in a peer which is not attached to the cluster. This would likely make things explode.
+
+
+## Challenges and open questions
+
+### The IPFS-like cluster API
+
+* Does it need to be and behave exactly like an IPFS node would ?
+* Is the IPFS proxy setup the right way to do composition?
+* Can the overcluster have any awareness that it is talking to an undercluster?
+* Can the proxy requests be forwarded via RPC if we are aware that we are attached not to an IPFS deaemon but to a cluster peer?
+* What are the latencies introduced by multi-level cluster, do they need to be better accounted in the IPFS connector?
+
+### Security
+
+* What authentication methods are needed in a composite setup? What authentication should the ipfs proxy support?
+* What is the trust relationship between the overclusters and the underclusters?
+* What authorization strategies make sense?
+  * Read only use?
+  * Only some enabled endpoints for some users?
+* Should channels be encrypted ?
+
+### Administration overhead
+
+* How hard is it to manage a composite cluster vs. managing a single one? Is it worth it to the use-cases?
```

Comment:
👍  This is going to be key to figure out when planning which features to tackle with composite approaches.